core/core.go

```diff
@@ -627,10 +639,25 @@ func (c *Core) WriteBlock(block *types.Block) {
 	if nodeCtx == common.PRIME_CTX {
 		if block != nil {
 			c.SetSyncTarget(block.Header())
+			if c.shouldStartSnapSync(block) {
+				log.Global.Debugf("Starting snap sync for block %d", block.NumberU64(c.NodeCtx()))
+				c.TriggerSnapSync(block.Header())
+			}
 		}
 	}
 }
 
+// Returns true if the chain tip is 500 blocks ahead
+// of the current block
+func (c *Core) shouldStartSnapSync(block *types.Block) bool {
+	// 1. Get the block from the chain tip
+	latest := c.CurrentBlock()
+
+	// 2. If latest block is more than 500 blocks ahead of the current block
+	// then start snap sync
+	return block.NumberU64(c.NodeCtx())+c_snapSyncTriggerThreshold < latest.NumberU64(c.NodeCtx())
```

Comment:
@gameofpointers how do we intend to determine when snap sync is necessary? IIUC, this is a security vulnerability. All I have to do is mine 1 fake block 1000000000000 blocks into the future, and you will ask to snapsync to me, at which point I can feed you any state db I wish, and you have no way to invalidate it.